### PR TITLE
fix: detect non-recursive CTEs in the recursive `WITH` clause

### LINKED ISSUE #9804: `WITH RECURSIVE` clause may define non-recursive CTEs
### Describe the bug

DataFusion did not recognize this scenario and treated them all as recursive CTEs, which led to incorrect results.

This is a reasonable usage. Users can define multiple CTEs using a WITH clause, one of them might be recursive, while others are not.
```sql
WITH RECURSIVE
non_recursive_cte AS (
  SELECT 1
),
recursive_cte AS (
  SELECT 1 AS a UNION ALL SELECT a+2 FROM recursive_cte WHERE a < 3
)
SELECT * FROM non_recursive_cte, recursive_cte;
```

### To Reproduce

Run the following query in CLI:
```sh
DataFusion CLI v36.0.0
❯ WITH RECURSIVE cte AS (
    SELECT a FROM (VALUES(1)) AS t(a) WHERE a > 2
    UNION ALL
    SELECT 2
) SELECT * FROM cte;
0 rows in set. Query took 0.016 seconds.
```

### Expected behavior

The result should be similar to PostgreSQL, with one row instead of zero rows.
```sh
psql=> WITH RECURSIVE cte AS (
    SELECT a FROM (VALUES(1)) AS t(a) WHERE a > 2
    UNION ALL
    SELECT 2
) SELECT * FROM cte;
 a
---
 2
(1 row)
```

### Additional context

The following is documentation from BigQuery. https://cloud.google.com/bigquery/docs/reference/standard-sql/query-syntax#recursive_keyword

A `WITH` clause can optionally include the `RECURSIVE` keyword, which does two things:
- Enables recursion in the `WITH` clause. If this keyword is not present, you can only include non-recursive common table expressions (CTEs). If this keyword is present, you can use **both recursive and non-recursive CTEs**.
- Changes the visibility of CTEs in the `WITH` clause. If this keyword is not present, a CTE is only visible to CTEs defined after it in the `WITH` clause. If this keyword is present, a CTE is visible to all CTEs in the `WITH` clause where it was defined.

### PR DESCRIPTION
**Pull request by jonahgao**
## Which issue does this PR close?

Closes #9804.

## Rationale for this change
If the recursive term of a CTE does not reference the CTE itself, then it should be compiled to a non-recursive CTE, otherwise it may lead to incorrect query results.

This PR traverses the `LogicalPlan` of the recursive term, looking for a table scan of the work table to determine if there is a self-reference.

## What changes are included in this PR?
- Move the CTE-related code to a new mod named `cte`.
- Detect non-recursive CTEs in the recursive `WITH` clause.

## Are these changes tested?
Yes

## Are there any user-facing changes?
No

### CODE REVIEW

File: datafusion/sql/src/cte.rs [hunk 1]
```diff
@@ -0,0 +1,212 @@
...
+        planner_context.insert_cte(cte_name.clone(), work_table_plan);
+
+        // ---------- Step 3: Compile the recursive term ------------------
+        // this uses the named_relation we inserted above to resolve the
+        // relation. This ensures that the recursive term uses the named relation logical plan
+        // and thus the 'continuance' physical plan as its input and source
+        let recursive_plan =
+            self.set_expr_to_plan(*right_expr, &mut planner_context.clone())?;
+
+        // Check if the recursive term references the CTE itself,
```
Comment (jonahgao) [inline datafusion/sql/src/cte.rs:172]: Here are the main changes.

File: datafusion/sql/src/set_expr.rs [hunk 2]
```diff
@@ -35,45 +35,58 @@ impl<'a, S: ContextProvider> SqlToRel<'a, S> {
...
+            SetQuantifier::AllByName => {
+                not_impl_err!("UNION ALL BY NAME not implemented")
+            }
+            SetQuantifier::DistinctByName => {
+                not_impl_err!("UNION DISTINCT BY NAME not implemented")
+            }
+        }
+    }
+
+    pub(super) fn set_operation_to_plan(
```
Comment (jonahgao) [inline datafusion/sql/src/set_expr.rs:63]: This function is reused by CTE.

File: datafusion/sql/tests/sql_integration.rs [hunk 3]
```diff
@@ -2992,16 +2992,6 @@ fn join_with_aliases() {
     quick_test(sql, expected);
 }
 
-#[test]
-fn cte_use_same_name_multiple_times() {
```
Comment (jonahgao) [inline datafusion/sql/tests/sql_integration.rs:2996]: This test is moved to slt.

**Review by alamb — APPROVED** [review]
Thank you @jonahgao  -- this PR looks really nice 🏆  -- I especially like how you have refactored the code into smaller and easier to understand functions

I apologize for the delay in reviewing.  🚀 

cc @matthewgapp

**Comment by jonahgao** [comment]
No worries about the delay :) And thank you for the review @alamb !

**Pull request merged**